service/matching/matchingEngine.go

```diff
@@ -253,11 +255,14 @@ func (e *matchingEngineImpl) updateTaskList(taskList *taskListID, mgr taskListMa
 	e.taskLists[*taskList] = mgr
 }
 
-func (e *matchingEngineImpl) removeTaskListManager(id *taskListID) {
+func (e *matchingEngineImpl) removeTaskListManager(tlMgr taskListManager) {
+	id := tlMgr.TaskListID()
 	e.taskListsLock.Lock()
-	defer e.taskListsLock.Unlock()
-
-	delete(e.taskLists, *id)
+	currentTlMgr, ok := e.taskLists[*id]
+	if ok && tlMgr == currentTlMgr {
```

Comment:
I would add a comment as to why your'e checking both ID and instance referencing the description at https://github.com/temporalio/temporal/pull/1917, since honestly it's a little confusing without looking at the PR and surrounding comment